crates/bevy_ecs/src/schedule/schedule.rs

```diff
@@ -661,36 +657,159 @@ impl ScheduleGraph {
+    }
+
+    fn configure_sets_inner(
+        &mut self,
+        configs: SystemSetConfigs,
+        ancestor_chained: bool,
+    ) -> NodesResult {
+        match configs {
+            SystemSetConfigs::SystemSetConfig(config) => {
+                let node_id = self.configure_set_inner(config).unwrap();
+                if ancestor_chained {
+                    NodesResult {
+                        densely_chained: true,
+                        nodes: vec![node_id],
+                    }
+                } else {
+                    NodesResult {
+                        densely_chained: true,
+                        nodes: Vec::new(),
+                    }
+                }
             }
-        } else {
-            for set in set_iter {
-                self.configure_set_inner(set).unwrap();
+            SystemSetConfigs::Configs {
+                mut configs,
+                collective_conditions,
+                chained,
+            } => {
+                let more_than_one_entry = configs.len() > 1;
+                if !collective_conditions.is_empty() {
+                    if more_than_one_entry {
+                        let set = AnonymousSet::new();
+                        for config in &mut configs {
+                            config.in_set_inner(set.dyn_clone());
+                        }
+                        let mut set_config = SystemSetConfig::new(set.dyn_clone());
+                        set_config.conditions.extend(collective_conditions);
+                        self.configure_set_inner(set_config).unwrap();
+                    } else {
+                        for condition in collective_conditions {
+                            configs[0].run_if_inner(condition);
+                        }
+                    }
+                }
+                let mut config_iter = configs.into_iter();
+                let mut nodes_in_scope = Vec::new();
+                let mut densely_chained = true;
+                if chained {
+                    let Some(prev) = config_iter.next() else {
+                        return NodesResult {
+                            nodes: Vec::new(),
+                            densely_chained: true
+                        }
+                    };
+                    let mut previous_result = self.configure_sets_inner(prev, true);
+                    densely_chained = previous_result.densely_chained;
+                    for current in config_iter {
+                        let current_result = self.configure_sets_inner(current, true);
+                        densely_chained = densely_chained && current_result.densely_chained;
+                        match (
+                            previous_result.densely_chained,
+                            current_result.densely_chained,
+                        ) {
+                            // Both groups are "densely" chained, so we can simplify the graph by only
```

Comment:
This is a very helpful comment, but more introduction of the concept of "densely chained" would be helpful early in this function.

Reply:
I previously just copied and adjusted the `add_systems_inner` function, but forgot to copy the documentation of `add_systems_inner` for `configure_sets_inner`, which explained what "densely chained" means. I refactored this PR to instead introduce generic `ItemConfig` and `ItemConfigs` types and a generic `process_configs` (a generic version of `add_systems_inner` and `configure_sets_inner`) to remove code duplication and make reviewing easier.